aospy/utils/times.py

```diff
@@ -571,3 +503,89 @@ def ensure_time_as_index(ds):
             da[TIME_STR] = ds[TIME_STR]
             ds[name] = da
     return ds
+
+
+def infer_year(date):
+    """Given a datetime-like object or string infer the year
+
+    Parameters
+    ----------
+    date : str
+        Input date
+
+    Returns
+    -------
+    int
+    """
+    if isinstance(date, str):
+        pattern = '(?P<year>\d{4})'
+        result = re.match(pattern, date)
+        if result:
+            return int(result.groupdict()['year'])
+        else:
+            raise ValueError('Invalid date string provided: {}'.format(date))
+    elif isinstance(date, np.datetime64):
+        return date.item().year
+    else:
+        return date.year
+    
+
+def maybe_cast_as_timestamp(date):
+    """Convert a date to a pd.Timestamp object if possible
+
+    Parameters
+    ----------
+    date : datetime-like object
+        Input datetime
+
+    Returns
+    -------
+    pd.Timestamp or cftime.datetime
+    """
+    if isinstance(date, str):
+        return np.datetime64(date).item()
+    else:
+        try:
+            return pd.to_datetime(date)
+        except (TypeError, OutOfBoundsDatetime):
+            return date
+
+
+def maybe_convert_to_index_date_type(index, date):
+    """Convert a datetime-like object to the appropriate type
```

Comment:
Nit: Here and elsewhere, let's end all the one-line summaries in the docstrings with a period.  I think it's a little cleaner, and it's more consistent with other packages docs.  I realize we probably haven't been 100% consistent on this in the past.

Reply:
The meaning of "appropriate type" isn't clear in this docstring...one or two sentences expanding this would help in cases like this one where a developer is returning to this after a while or is new to it.